### ISSUE TITLE
WhisperKit: SuppressTokensFilter writes to read-only MLMultiArray when suppressTokens includes -1 on iOS 26

### ISSUE THREAD

**Issue by 7kylor** [argmaxinc/WhisperKit#392]
## Bug Report: EXC_BAD_ACCESS crash in SuppressTokensFilter.filterLogits on iOS 26

I'm working on an iOS local AI models app and got this report crash from TestFlight [Test it here](https://testflight.apple.com/join/ntUKmRpq) 
Debugged it and reported this for a fix as applicable. 

### Environment
- **WhisperKit Version**: main branch (commit 9c673e35193fe4e67593dd2973c157ebe666e598)
- **Device**: iPhone 16 Pro (iPhone18,1)
- **iOS Version**: 26.2 (23C55)
- **Xcode Version**: 17C53

### Description
App crashes with `EXC_BAD_ACCESS (SIGBUS)` when using `AudioStreamTranscriber` for streaming transcription. The crash occurs in `SuppressTokensFilter.filterLogits` when attempting to write to a read-only `MLMultiArray`.

### Crash Details
- Thread 9 (crashed)***
- Location: MLMultiArray.fill<A>(indexes:with:) at Extensions+Public.swift:126
- Called from: SuppressTokensFilter.filterLogits(_:withTokens:) at LogitsFilter.swift:23
- Error: EXC_BAD_ACCESS (SIGBUS) with KERN_PROTECTION_FAILURE - attempting to write to read-only memory
- Exception Type: EXC_BAD_ACCESS (SIGBUS) 
- Exception Subtype: KERN_PROTECTION_FAILURE at 0x000000012a01fffe
- Termination Reason: SIGNAL 10 Bus error: 10
- esr: 0x9200004f (Data Abort) byte write Permission fault

> Thread 9 Crashed:
> 0 MLMultiArray.fill<A>(indexes:with:) + 432 (Extensions+Public.swift:126)
> 1 SuppressTokensFilter.filterLogits(:withTokens:) + 20 (LogitsFilter.swift:23)
> 2 protocol witness for LogitsFiltering.filterLogits(:withTokens:) in conformance SuppressTokensFilter
> 3 TextDecoder.decodeText(from:using:sampler:options:callback:) + 356 (TextDecoder.swift:831)
> 4 decodeWithFallback in TranscribeTask.run(audioArray:decodeOptions:callback:) (TranscribeTask.swift:338)
> 5 TranscribeTask.run(audioArray:decodeOptions:callback:) (TranscribeTask.swift:175)
> 6 AudioStreamTranscriber.transcribeAudioSamples(:) (AudioStreamTranscriber.swift:196)
> 7 AudioStreamTranscriber.transcribeCurrentBuffer() (AudioStreamTranscriber.swift:159)
> 8 AudioStreamTranscriber.startStreamTranscription() (AudioStreamTranscriber.swift:98)
> 

### Root Cause Analysis
On iOS 26, CoreML appears to return read-only `MLMultiArray` instances for performance optimization. The `SuppressTokensFilter.filterLogits` method attempts to modify the logits array in-place using `MLMultiArray.fill(indexes:with:)`, which triggers a memory protection fault.

### DecodingOptions Used
```
DecodingOptions(
    task: .transcribe,
    language: nil,  // auto-detect
    temperature: 0.2,
    temperatureFallbackCount: 3,
    sampleLength: 224,
    usePrefillPrompt: false,
    usePrefillCache: false,
    skipSpecialTokens: true,
    withoutTimestamps: true,
    wordTimestamps: false,
    suppressBlank: true,
    supressTokens: [-1],  // Default suppression - THIS TRIGGERS THE CRASH
    compressionRatioThreshold: 2.4,
    logProbThreshold: -1.0,
    firstTokenLogProbThreshold: -1.5,
    noSpeechThreshold: 0.4
) 
```

### Workaround
Setting supressTokens: [] (empty array) prevents the crash by bypassing the SuppressTokensFilter code path entirely for the time being (let me know if there's a batter way to fix this).

### Suggested Fix
The MLMultiArray.fill operation in Extensions+Public.swift should create a mutable copy of the array before attempting to modify it, or check if the array is writable before attempting in-place modification.

```
`// Potential fix in Extensions+Public.swift
// Before modifying, create a mutable copy if the original is read-only
let mutableLogits = try? MLMultiArray(shape: logits.shape, dataType: logits.dataType)
// Copy data and use mutableLogits instead`
```
 
 
### Steps to Reproduce
1- Use AudioStreamTranscriber for streaming transcription on iOS 26 (used it locally [Test it here](https://testflight.apple.com/join/ntUKmRpq) )
2- Set supressTokens: [-1] in DecodingOptions
3- Start streaming transcription
4- App crashes when filterLogits is called

### Additional Context
The crash only occurs on iOS 26+ where CoreML returns read-only arrays
Non-streaming transcription may also be affected
The VM Region Info shows the crash address is in a gap between a mapped file (read-only) and a malloc region

**Comment by atiorh** [comment]
Thank you for the report! We have reproduced this and will come back to this issue shortly.

**Comment by ZachNagengast** [comment]
@7kylor Thanks for the report, do you have a specific use case for [-1] supress tokens? Or is it crashing for other input values? We should definitely sanitize this input but wondering if its coming up in other cases besides an invalid input.

**Comment by 7kylor** [comment]
Thanks for the quick response @atiorh @ZachNagengast 

I had a crash report from a TestFlight user. I set the value of supressTokens: [-1] which  is not an invalid input it works for me - I'm using it as documented to enable default token suppression. The crash is not specific to the -1 value. The root cause is that on iOS 26 with CoreML returns read-only MLMultiArray instances for the logits which the WhisperKit isn't using yet. Any attempt to modify this array in-place will crash.

Tracing the stack at: 
```
MLMultiArray.fill<A>(indexes:with:) + 432 (Extensions+Public.swift:126)
SuppressTokensFilter.filterLogits(_:withTokens:) + 20 (LogitsFilter.swift:23)
```

When looking deeper at the filterLogits method it attempts to write to the logits array (setting suppressed token positions to -Float.infinity). This write operation triggers the KERN_PROTECTION_FAILURE because the array is read-only. So It'd crash for any non-empty supressTokens array - whether it's [-1] or any other valid token ID(s) because the underlying issue is the write operation not the specific values.

The workaround I'm currently using supressTokens: [] which works because it bypasses the SuppressTokensFilter entirely (no tokens to suppress = no write operation attempted). And this may output odd characters or tokens in output i suppose (correct me if I'm wrong). 

I suggest to create a mutable copy of the logits array before modifying it in filterLogits.

**Comment by ZachNagengast** [comment]
Got it thanks for the context. This repo doesnt use the default set of suppress tokens at the moment so -1 doesn't get replaced. When [-1] is passed in we will have a -1 at some point in this `linearOffset`

https://github.com/argmaxinc/WhisperKit/blob/9c673e35193fe4e67593dd2973c157ebe666e598/Sources/WhisperKit/Utilities/Extensions%2BPublic.swift#L124-L127

Since this is an unsafe pointer, this will lead to some undefined behavior by attempting to write outside the intended buffer. Importantly, this wont _always_ crash, since unsafe pointers specifically allow that kind of thing. I think this may be where you KERN_PROTECTION_FAILURE error is coming from - if this is true, it's actually not attempting to write to the intended MLMultiArray, but some unknown memory region very close to it, which could be read only in this case. You might notice the crash doesnt occur immediately on every token, but then occurs after some amount of time, which would be when it happend to hit a read only portion of memory in that -1 location.

You mentioned "Any attempt to modify this array in-place will crash." so I'd be interested if you were able to reproduce it without the [-1] input, but your workaround seems to be a good option in the meantime.

**Comment by 7kylor** [comment]
Thanks for the detailed explanation. I re-analyze the crash report and your diagnosis is more accurate. 

```
 Exception Subtype: KERN_PROTECTION_FAILURE at 0x000000012a01fffe
VM Region Info: 0x12a01fffe is not in any region.

REGION TYPE                 START - END      [ VSIZE] PRT/MAX SHRMOD
mapped file              122f98000-1255f0000 [ 38.3M] r--/r-- SM=PRV
--->  GAP OF 0x4e10000 BYTES (~82MB gap)
MALLOC_LARGE             12a400000-12ae00000 [ 10.0M] rw-/rwx SM=PRV
```

The crash address 0x12a01fffe is in an unmapped gap between regions, not in the read-only mapped file region. If it were a read-only array issue, the crash would be within the mapped file region at 0x122f98000-0x1255f0000. Using the -1 as a literal index with stride calculations produces a large negative offset that lands in unmapped memory causing the bus error. My workaround with supressTokens: [] resolves it. 

I haven't been able to reproduce it - I went directly to supressTokens: [] as a workaround after the initial crash, since it happened on an external tester device. My initial analysis was based on the KERN_PROTECTION_FAILURE error, but your explanation about the -1 index causing an out-of-bounds write makes sense given that the crash address is in unmapped memory (a gap between regions), not within a read-only region.
